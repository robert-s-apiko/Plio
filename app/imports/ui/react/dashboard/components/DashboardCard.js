import PropTypes from 'prop-types';
import React from 'react';
import { Card, CardBody, CardTitle, CardFooter, CardImg } from 'reactstrap';
import cx from 'classnames';
import styled from 'styled-components';

import { CardBlock, Icon } from '../../components';

const Wrapper = styled.div`
  min-height: 92px;
`;

const DashboardCard = ({
  className,
  disabled,
  title,
  footer,
  icon,
  text,
  label,
  ...props
}) => (
  <Card
    tag="a"
    className={cx(
      className,
      'dashboard-item',
      { 'dashboard-item-disabled': disabled },
    )}
    {...props}
  >
    <CardBlock>
      <Wrapper>
        <CardImg
          top
          tag={Icon}
          className="dashboard-item-icon"
          name={cx('fw', icon)}
        />
        <CardBody>
          <CardTitle>
            {title}
          </CardTitle>
        </CardBody>
      </Wrapper>
      <CardFooter>
        <span className="text-muted">{text}</span>
        {label && (<span className="label label-danger margin-left">{label}</span>)}
      </CardFooter>
    </CardBlock>
  </Card>
);

DashboardCard.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  icon: PropTypes.string,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  text: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  footer: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
};

export default DashboardCard;
