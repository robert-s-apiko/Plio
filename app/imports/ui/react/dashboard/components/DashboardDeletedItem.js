import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { compose, withProps, onlyUpdateForKeys, withHandlers } from 'recompose';
import { Button, ButtonGroup } from 'reactstrap';
import { getFormattedDate } from '../../../../share/helpers';
import Label from '../../components/Labels/Label';

const StyledDeletedGoal = styled.div`
  cursor: inherit;
  .btn-group {
    position: relative;
    top: -2px;
    float: right;
    .btn.btn-secondary:first-child:not(:last-child) {
      padding-right: .75rem;
    }
  }
`;

const enhance = compose(
  onlyUpdateForKeys(['_id', 'isDeleted', 'completedAt', 'deletedAt', 'title', 'sequentialId']),
  withHandlers({
    onRestore: ({
      _id,
      isDeleted,
      onUndoCompletion,
      onRestore,
    }) => () => isDeleted ? onRestore(_id) : onUndoCompletion(_id),
    onRemove: ({ _id, title, onRemove }) => () => onRemove({ _id, title }),
  }),
  withProps(({ canRestore, ...props }) => ({
    date: props.isDeleted ? props.deletedAt : props.completedAt,
    statusText: props.isDeleted ? 'deleted' : 'completed',
    canRestore: canRestore(props),
  })),
);

const DashboardDeletedItem = ({
  sequentialId,
  title,
  date,
  statusText,
  onRestore,
  onRemove,
  canRestore,
}) => (
  <StyledDeletedGoal className="dashboard-stats-action">
    <Label>{sequentialId}</Label>
    {title}
    <span className="text-muted"> ({statusText} {getFormattedDate(date)})</span>
    <ButtonGroup size="sm">
      <Button onClick={onRemove}>Delete</Button>
      <Button disabled={!canRestore} onClick={onRestore}>Restore</Button>
    </ButtonGroup>
  </StyledDeletedGoal>
);

DashboardDeletedItem.propTypes = {
  sequentialId: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  statusText: PropTypes.string.isRequired,
  date: PropTypes.number.isRequired,
  onRestore: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired,
  canRestore: PropTypes.bool.isRequired,
};

export default enhance(DashboardDeletedItem);
