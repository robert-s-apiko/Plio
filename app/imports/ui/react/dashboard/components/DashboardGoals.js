import PropTypes from 'prop-types';
import React from 'react';
import pluralize from 'pluralize';
import { joinIds } from 'plio-util';

import {
  DashboardStatsExpandable,
  IconLoading,
  PlusButton,
  GoalsChart,
} from '../../components';
import { GoalAddModalContainer, GoalEditModalContainer } from '../../goals/containers';

const DashboardGoals = ({
  totalCount,
  goals,
  onLineTap,
  toggle,
  isOpen,
  loading,
  isAddModalOpen,
  toggleAddModal,
  userId,
  organizationId,
  isEditModalOpen,
  toggleEditModal,
}) => (
  <DashboardStatsExpandable
    items={goals}
    total={totalCount}
    itemsPerRow={goals.length}
    renderIcon={loading ? () => <IconLoading /> : undefined}
    render={({ items }) => !!items.length && (
      <GoalsChart
        key={joinIds(items)}
        goals={items}
        onScatterTap={toggleEditModal}
        {...{ onLineTap }}
      />
    )}
    {...{ toggle, isOpen }}
  >
    <PlusButton size="1" onClick={toggleAddModal} />
    {pluralize('goal', totalCount || goals.length, true)}
    <GoalAddModalContainer
      isOpen={isAddModalOpen}
      toggle={toggleAddModal}
      ownerId={userId}
      {...{ organizationId }}
    />
    <GoalEditModalContainer
      isOpen={isEditModalOpen}
      toggle={toggleEditModal}
      {...{ organizationId }}
    />
  </DashboardStatsExpandable>
);

DashboardGoals.propTypes = {
  totalCount: PropTypes.number.isRequired,
  goals: PropTypes.arrayOf(PropTypes.object).isRequired,
  // zoomDomain: ?
  onZoom: PropTypes.func,
  onLineTap: PropTypes.func,
  onScatterTap: PropTypes.func,
  toggle: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  isAddModalOpen: PropTypes.bool.isRequired,
  toggleAddModal: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  userId: PropTypes.string.isRequired,
  organizationId: PropTypes.string.isRequired,
  isEditModalOpen: PropTypes.bool.isRequired,
  toggleEditModal: PropTypes.func.isRequired,
};

export default DashboardGoals;
