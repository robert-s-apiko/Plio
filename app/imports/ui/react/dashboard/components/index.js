export { default as DashboardUserStats } from './DashboardUserStats';
export { default as DashboardStatsMessage } from './DashboardStatsMessage';
export { default as DashboardStatsOverdueItems } from './DashboardStatsOverdueItems';
export { default as DashboardStatsUnreadMessages } from './DashboardStatsUnreadMessages';
export { default as DashboardStatsUserList } from './DashboardStatsUserList';
export { default as DashboardStatsAction } from './DashboardStatsAction';
export { default as DashboardGoals } from './DashboardGoals';
