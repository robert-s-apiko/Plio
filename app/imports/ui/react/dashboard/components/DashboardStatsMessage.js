import PropTypes from 'prop-types';
import React from 'react';

import { Icon } from '../../components';

const DashboardStatsMessage = ({
  url,
  extension,
  fullName,
  text,
  timeString,
}) => (
  <a
    className="dashboard-stats-message"
    href={url}
  >
    {extension ? (
      <Icon name={`file-${extension}-o`} />
    ) : (
      <Icon name="comment" />
    )}
    <strong>{fullName}: </strong>
    {text}
    <span className="text-muted">{timeString} ago</span>
  </a>
);

DashboardStatsMessage.propTypes = {
  url: PropTypes.string.isRequired,
  fullName: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  timeString: PropTypes.string.isRequired,
  extension: PropTypes.string,
};

export default DashboardStatsMessage;
