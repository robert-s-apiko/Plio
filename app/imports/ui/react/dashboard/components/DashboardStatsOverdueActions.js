import PropTypes from 'prop-types';
import React from 'react';
import pluralize from 'pluralize';
import { map } from 'ramda';
import { onlyUpdateForKeys } from 'recompose';

import { DashboardStatsExpandable } from '../../components';
import {
  WorkspaceDefaults,
  WorkspaceDefaultsTypes,
} from '../../../../share/constants';
import { DashboardStatsActionContainer } from '../containers';
import { joinIds } from '../../../../client/util';

const enhance = onlyUpdateForKeys(['count', 'workItems', 'orgSerialNumber', 'itemsPerRow']);

export const DashboardStatsOverdueActions = ({
  count,
  workItems,
  orgSerialNumber,
  itemsPerRow = WorkspaceDefaults[WorkspaceDefaultsTypes.DISPLAY_MESSAGES],
}) => (
  <DashboardStatsExpandable
    items={workItems}
    itemsPerRow={itemsPerRow}
    render={({ items }) => (
      <div key={joinIds(items)}>
        {map(item => (
          <DashboardStatsActionContainer key={item._id} {...{ ...item, orgSerialNumber }} />
        ), workItems)}
      </div>
    )}
  >
    {pluralize('overdue work item', count, true)}
  </DashboardStatsExpandable>
);

DashboardStatsOverdueActions.propTypes = {
  workItems: PropTypes.arrayOf(PropTypes.object).isRequired,
  itemsPerRow: PropTypes.number.isRequired,
  orgSerialNumber: PropTypes.number.isRequired,
  count: PropTypes.number.isRequired,
};

export default enhance(DashboardStatsOverdueActions);
