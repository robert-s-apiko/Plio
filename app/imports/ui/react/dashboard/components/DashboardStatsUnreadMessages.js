import PropTypes from 'prop-types';
import React from 'react';
import pluralize from 'pluralize';

import { DashboardStats } from '../../components';
import DashboardStatsUserList from './DashboardStatsUserList';

export const DashboardUserStats = ({
  messages,
  count,
}) => !!messages.length && (
  <DashboardStats>
    <DashboardStats.Title>
      {count}
      
    </DashboardStats.Title>
  </DashboardStats>
);

DashboardUserStats.propTypes = {

};

export default DashboardUserStats;
