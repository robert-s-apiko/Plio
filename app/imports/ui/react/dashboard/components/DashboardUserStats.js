import PropTypes from 'prop-types';
import React from 'react';
import pluralize from 'pluralize';
import { branch } from 'recompose';
import { converge, gt, prop, identity } from 'ramda';

import { getUsersLength } from '../../../../client/util';
import { DashboardStats, Collapse, Button, Icon } from '../../components';
import DashboardStatsUserList from './DashboardStatsUserList';
import { withStateToggle, namedCompose } from '../../helpers';

const usersExceedLimit = converge(gt, [
  getUsersLength,
  prop('usersPerRow'),
]);

const enhance = namedCompose('DashboardUserStats')(
  branch(
    usersExceedLimit,
    withStateToggle(false, 'isOpen', 'toggle'),
    identity,
  ),
);

const DashboardUserStats = enhance(({
  users,
  isOpen,
  toggle,
  usersPerRow,
}) => !!users.length && (
  <DashboardStats>
    <DashboardStats.Title>
      {!!toggle && (
        <Button size="1" color="secondary add" onClick={toggle}>
          <Icon name="plus" />
        </Button>
      )}
      {pluralize('user', users.length, true)} online
    </DashboardStats.Title>
    <DashboardStatsUserList users={users.slice(0, usersPerRow)} />

    {!!toggle && (
      <Collapse {...{ isOpen }}>
        <DashboardStatsUserList users={users.slice(usersPerRow)} />
      </Collapse>
    )}
  </DashboardStats>
));

DashboardUserStats.propTypes = {
  users: PropTypes.arrayOf(PropTypes.object).isRequired,
  usersPerRow: PropTypes.number.isRequired,
};

export default DashboardUserStats;
