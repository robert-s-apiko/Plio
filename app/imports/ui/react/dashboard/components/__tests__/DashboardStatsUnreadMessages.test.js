import React from 'react';
import { shallow } from 'enzyme';
import { times } from 'ramda';

import DashboardStatsUnreadMessages from '../DashboardStatsUnreadMessages';
import { DashboardStatsMessageContainer } from '../../containers';
import { Button } from '../../../components';

describe('DashboardStatsUnreadMessages', () => {
  const realDate = Date;

  beforeEach(() => {
    global.Date = jest.fn();
  });
  afterEach(() => {
    global.Date = realDate;
  });

  const markAllAsRead = jest.fn();
  const loadAll = jest.fn();
  const loadLimited = jest.fn();
  const orgSerialNumber = 1;
  const genMessage = n => ({
    _id: n,
    url: `http://example.com/${n}`,
    extension: n % 2 === 0 ? 'docx' : undefined,
    fullName: `John Doe ${n}`,
    text: `Hello World #${n}`,
    timeString: (new Date()).toString(),
  });
  it('renders messages correctly & does not render show/hide button', () => {
    const len = 24;
    const messages = times(genMessage, len);

    const wrapper = shallow(<DashboardStatsUnreadMessages
      {...{
        messages,
        markAllAsRead,
        loadAll,
        loadLimited,
        orgSerialNumber,
        count: len,
      }}
    />);

    expect(wrapper.find(DashboardStatsMessageContainer)).toHaveLength(len);
  });

  // eslint-disable-next-line max-len
  it('renders hide/show button only if total count of unread messages is greater than current', () => {
    const getProps = len => ({
      count: 24,
      markAllAsRead,
      loadAll,
      loadLimited,
      orgSerialNumber,
      isLimitEnabled: true,
      messages: times(genMessage, len),
    });

    const wrapper1 = shallow(<DashboardStatsUnreadMessages {...getProps(24)} />);
    const wrapper2 = shallow(<DashboardStatsUnreadMessages {...getProps(15)} />);

    expect(wrapper1.find(Button)).toHaveLength(0);
    expect(wrapper2.find(Button)).toHaveLength(1);
  });

  it('matches snapshot', () => {
    const len = 24;
    const messages = times(genMessage, len);
    const wrapper = shallow(<DashboardStatsUnreadMessages
      {...{
        messages,
        markAllAsRead,
        loadAll,
        loadLimited,
        orgSerialNumber,
        count: len,
      }}
    />);

    expect(wrapper).toMatchSnapshot();
  });
});
