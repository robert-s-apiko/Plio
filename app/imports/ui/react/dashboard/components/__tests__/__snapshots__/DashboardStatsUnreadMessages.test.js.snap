// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardStatsUnreadMessages matches snapshot 1`] = `
<DashboardStats>
  <DashboardStatsTitle>
    24 unread messages
    <a
      className="pointer"
      onClick={[Function]}
      title="Mark all messages as read"
    >
      <Icon
        name="times-circle"
      />
    </a>
  </DashboardStatsTitle>
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={0}
    extension="docx"
    fullName="John Doe 0"
    key="0"
    orgSerialNumber={1}
    text="Hello World #0"
    timeString="[object Object]"
    url="http://example.com/0"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={1}
    fullName="John Doe 1"
    key="1"
    orgSerialNumber={1}
    text="Hello World #1"
    timeString="[object Object]"
    url="http://example.com/1"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={2}
    extension="docx"
    fullName="John Doe 2"
    key="2"
    orgSerialNumber={1}
    text="Hello World #2"
    timeString="[object Object]"
    url="http://example.com/2"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={3}
    fullName="John Doe 3"
    key="3"
    orgSerialNumber={1}
    text="Hello World #3"
    timeString="[object Object]"
    url="http://example.com/3"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={4}
    extension="docx"
    fullName="John Doe 4"
    key="4"
    orgSerialNumber={1}
    text="Hello World #4"
    timeString="[object Object]"
    url="http://example.com/4"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={5}
    fullName="John Doe 5"
    key="5"
    orgSerialNumber={1}
    text="Hello World #5"
    timeString="[object Object]"
    url="http://example.com/5"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={6}
    extension="docx"
    fullName="John Doe 6"
    key="6"
    orgSerialNumber={1}
    text="Hello World #6"
    timeString="[object Object]"
    url="http://example.com/6"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={7}
    fullName="John Doe 7"
    key="7"
    orgSerialNumber={1}
    text="Hello World #7"
    timeString="[object Object]"
    url="http://example.com/7"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={8}
    extension="docx"
    fullName="John Doe 8"
    key="8"
    orgSerialNumber={1}
    text="Hello World #8"
    timeString="[object Object]"
    url="http://example.com/8"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={9}
    fullName="John Doe 9"
    key="9"
    orgSerialNumber={1}
    text="Hello World #9"
    timeString="[object Object]"
    url="http://example.com/9"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={10}
    extension="docx"
    fullName="John Doe 10"
    key="10"
    orgSerialNumber={1}
    text="Hello World #10"
    timeString="[object Object]"
    url="http://example.com/10"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={11}
    fullName="John Doe 11"
    key="11"
    orgSerialNumber={1}
    text="Hello World #11"
    timeString="[object Object]"
    url="http://example.com/11"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={12}
    extension="docx"
    fullName="John Doe 12"
    key="12"
    orgSerialNumber={1}
    text="Hello World #12"
    timeString="[object Object]"
    url="http://example.com/12"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={13}
    fullName="John Doe 13"
    key="13"
    orgSerialNumber={1}
    text="Hello World #13"
    timeString="[object Object]"
    url="http://example.com/13"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={14}
    extension="docx"
    fullName="John Doe 14"
    key="14"
    orgSerialNumber={1}
    text="Hello World #14"
    timeString="[object Object]"
    url="http://example.com/14"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={15}
    fullName="John Doe 15"
    key="15"
    orgSerialNumber={1}
    text="Hello World #15"
    timeString="[object Object]"
    url="http://example.com/15"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={16}
    extension="docx"
    fullName="John Doe 16"
    key="16"
    orgSerialNumber={1}
    text="Hello World #16"
    timeString="[object Object]"
    url="http://example.com/16"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={17}
    fullName="John Doe 17"
    key="17"
    orgSerialNumber={1}
    text="Hello World #17"
    timeString="[object Object]"
    url="http://example.com/17"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={18}
    extension="docx"
    fullName="John Doe 18"
    key="18"
    orgSerialNumber={1}
    text="Hello World #18"
    timeString="[object Object]"
    url="http://example.com/18"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={19}
    fullName="John Doe 19"
    key="19"
    orgSerialNumber={1}
    text="Hello World #19"
    timeString="[object Object]"
    url="http://example.com/19"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={20}
    extension="docx"
    fullName="John Doe 20"
    key="20"
    orgSerialNumber={1}
    text="Hello World #20"
    timeString="[object Object]"
    url="http://example.com/20"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={21}
    fullName="John Doe 21"
    key="21"
    orgSerialNumber={1}
    text="Hello World #21"
    timeString="[object Object]"
    url="http://example.com/21"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={22}
    extension="docx"
    fullName="John Doe 22"
    key="22"
    orgSerialNumber={1}
    text="Hello World #22"
    timeString="[object Object]"
    url="http://example.com/22"
  />
  <DashboardStatsMessageContainer(DashboardStatsMessage)
    _id={23}
    fullName="John Doe 23"
    key="23"
    orgSerialNumber={1}
    text="Hello World #23"
    timeString="[object Object]"
    url="http://example.com/23"
  />
  <defaultProps(mapProps(componentFromProp(component)))
    color="secondary"
    component="a"
    onClick={[Function]}
  >
    <span>
      Hide
    </span>
    <span>
      (
      23
       items)
    </span>
  </defaultProps(mapProps(componentFromProp(component)))>
</DashboardStats>
`;
